src/storage/src/chunk.rs

```diff
@@ -151,51 +154,84 @@ impl ChunkReaderBuilder {
         self
     }
 
-    pub async fn build(mut self) -> Result<ChunkReaderImpl> {
-        let time_range_predicate = self.build_time_range_predicate();
-        debug!(
-            "Time range predicate for chunk reader: {:?}",
-            time_range_predicate
-        );
+    /// Set time windows for scan.
+    pub fn time_windows(mut self, windows: Vec<TimestampRange>) -> Self {
+        self.time_windows = Some(windows);
+        self
+    }
 
-        let schema = Arc::new(
-            ProjectedSchema::new(self.schema, self.projection)
-                .context(error::InvalidProjectionSnafu)?,
-        );
+    pub async fn build_windowed(
+        self,
+        schema: &ProjectedSchemaRef,
+        time_range_predicate: &TimestampRange,
+        windows: Vec<TimestampRange>,
+    ) -> Result<BoxedBatchReader> {
+        let mut readers = Vec::with_capacity(windows.len());
+        for window in windows {
+            debug!("Scan time window: {:?}", window);
+            let time_range_predicate = time_range_predicate.and(&window);
+            let reader = self.build_reader(schema, &time_range_predicate).await?;
+            readers.push(reader);
+        }
+        let windowed_reader = WindowedReader::new(schema.clone(), readers);
+        Ok(Box::new(windowed_reader) as Box<_>)
+    }
 
+    async fn build_reader(
+        &self,
+        schema: &ProjectedSchemaRef,
+        time_range: &TimestampRange,
+    ) -> Result<BoxedBatchReader> {
         let num_sources = self.memtables.len() + self.files_to_read.len();
         let mut reader_builder = MergeReaderBuilder::with_capacity(schema.clone(), num_sources)
             .batch_size(self.iter_ctx.batch_size);
 
-        self.iter_ctx.projected_schema = Some(schema.clone());
-        for mem in self.memtables {
+        for mem in &self.memtables {
```

Comment:
The memtable might also need to support scan by time range.

Reply:
This PR depends on #1576 so that we can filter timestamps in memtable.